[sparse][ci] Add cuSPASRELt to CI

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115988
* __->__ #115369

Summary:

This PR adds in cuSPARSELt v0.4.07 into CI (12.1 and 11.8 CUDA) to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jcaip committed:
[sparse][ci] Add cuSPASRELt v0.4.0 to CI

Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

huydhn commented on .ci/docker/common/install_cusparselt.sh:
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
```
I think you could add `set -ex` here to harden the script and also allow the bash steps to be printed on GitHub console log to very that the installation works correctly

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented:
@pytorchbot rebase

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

malfet commented on .ci/docker/common/install_cusparselt.sh:
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+set -ex
+echo "this is a first test"
+
+if [ -n "${USE_CUSPARSELT}" ]; then
+    # cuSPARSELt license: https://docs.nvidia.com/cuda/cusparselt/license.html
+    mkdir tmp_cusparselt && cd tmp_cusparselt
+    # CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.5.0.1-archive.tar.xz"
+    if [[ ${CUDA_VERSION:0:4} == "12.1" ]]; then
+        CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.4.0.7-archive"
+        curl --retry 3 -OLs https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz
+    elif [[ ${CUDA_VERSION:0:4} == "11.8" ]]; then
+        CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.4.0.7-archive"
+        curl --retry 3 -OLs https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz
+    fi
+
+    tar xf ${CUSPARSELT_NAME}.tar.xz
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/local/cuda/include/
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/x86_64-linux-gnu/
+
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/local/cuda/lib64/
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/lib/x86_64-linux-gnu/
+    cd ..
+    rm -rf tmp_cusparselt
+    ldconfig
+fi
+
+echo "This is a test"
```
```suggestion
set -ex

# cuSPARSELt license: https://docs.nvidia.com/cuda/cusparselt/license.html
mkdir tmp_cusparselt
pushd tmp_cusparselt
CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.4.0.7-archive"
curl --retry 3 -OLs "https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz"
tar xf "${CUSPARSELT_NAME}.tar.xz"
cp -a "${CUSPARSELT_NAME}/include/*" /usr/local/cuda/include/
cp -a ${CUSPARSELT_NAME}/lib/* /usr/local/cuda/lib64/
popd
rm -rf tmp_cusparselt
ldconfig
```

malfet committed:
Apply Occam's razor to PR

Remove all unnecessary conditionals

jcaip commented on .ci/docker/common/install_cusparselt.sh:
```diff
@@ -0,0 +1,14 @@
+CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.4.0.7-archive"
+curl --retry 3 -OLs "https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz"
+tar xf "${CUSPARSELT_NAME}.tar.xz"
+cp -a "${CUSPARSELT_NAME}/include/*" /usr/local/cuda/include/
```
I think these quotes are breaking the script @malfet

jcaip committed:
fix quotes

jcaip commented:
@pytorchbot rebase

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt v0.4.0 to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.5.2.1 for cuda 12.1
v0.4.0.7 for cuda 11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.5.2.1 for cuda 12.1
v0.4.0.7 for cuda 11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip commented on test/test_sparse_semi_structured.py:
```diff
@@ -598,6 +598,8 @@ def setUp(self):
 
     @parametrize("dense_input_shape", [(128, 128)])
     def test_cslt_sparse_mm_int8_in_fp16_out(self, dense_input_shape, device):
+        assert False
```
just to make sure it's getting built with cusparselt, will remove later

malfet replied:
You were building with CUSPARELT, see https://github.com/pytorch/pytorch/actions/runs/7215814225/job/19660823130#step:12:1169

malfet reviewed (changes requested):
[Edit] Please note, that `rebase` undid all the changes I've made yesterday, please get rid of  `USE_CUSPARSELT`  unless you plan to add CI config without it.
Also, don't plan to test two different versions of cusparseLT, unless there is a good reason to do it

jcaip commented:
@malfet Sorry, I rebased to fix tests, didn't realize it would overwrite your changes.

0.5.2.1 requires >12.0 and 0.4.0.7 requires > 11.8, which is why I added both. Should I just default to 0.5.2.1 and restrict it to run on the 12.1 runners only?

malfet commented:
@jcaip can you share link to the doc that specifies min requirements for the cusparselt?

jcaip commented:
https://docs.nvidia.com/cuda/cusparselt/release_notes.html

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.5.2.1 for cuda 12.1
v0.4.0.7 for cuda 11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.5.2.1 for cuda 12.1
v0.4.0.7 for cuda 11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.5.2.1 for cuda 12.1
v0.4.0.7 for cuda 11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

malfet commented on .ci/docker/common/install_cusparselt.sh:
```diff
@@ -0,0 +1,21 @@
+    curl --retry 3 -OLs https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz
+
+    tar xf ${CUSPARSELT_NAME}.tar.xz
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/
```
```suggestion
```

malfet commented on .ci/docker/common/install_cusparselt.sh:
```diff
@@ -0,0 +1,21 @@
+    tar xf ${CUSPARSELT_NAME}.tar.xz
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/local/cuda/include/
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/x86_64-linux-gnu/
```
```suggestion
```

malfet commented on .ci/docker/common/install_cusparselt.sh:
```diff
@@ -0,0 +1,21 @@
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/x86_64-linux-gnu/
+
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/local/cuda/lib64/
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/lib/x86_64-linux-gnu/
```
```suggestion
```

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.5.2.1 for cuda 12.1
v0.4.0.7 for cuda 11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.5.2.1 for cuda 12.1
v0.4.0.7 for cuda 11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jcaip committed:
Update on "[sparse][ci] Add cuSPASRELt to CI"


Summary:

This PR adds in cuSPARSELt into CI to run our cuSPARSELt specific tests.

v0.4.0.7 for cuda 12.1/11.8

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -i

malfet commented:
@pytorchbot merge -f "Windows failures are irrelevant"
